### ISSUE TITLE
Explore GitHub Actions as alternative to Travis

### ISSUE THREAD

**Issue by kabalin** [moodlehq/moodle-plugin-ci#48]
As GitHub Actions is free for [public open source repos](https://github.com/features/actions#pricing-details), I suggest to consider making it a default CI option for plugin developers who use `moodle-plugin-ci` (also given GitHub is a popular choice among community plugin developers).

For Travis free usage withdraw discussion, please refer to #47.

Possible steps requred (some may require a separate issue to capture work):
- [x] Proof-of-concept how `moodle-plugin-ci` can be run using GitHub Actions (and support all test steps, docker, etc. no free tier limitations of tools that can be used)
- [x] Moving `moodle-plugin-ci` itself to GitHub Actions (test and release steps, updating release manual) (#54, #60)
- [x] Creating `gha.dist.yml` and adding documentation explaining its content and usage (#72)
- [ ] Adding instruction on wiki (similar to https://docs.moodle.org/dev/Travis_integration).

**Comment by kabalin** [comment]
https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/migrating-from-travis-ci-to-github-actions

**Comment by timhunt** [comment]
OK folks, I have sort-of made it work:

https://github.com/moodleou/moodle-filter_embedquestion/blob/github-actions/.github/workflows/ci.yml

generated this:

https://github.com/moodleou/moodle-filter_embedquestion/runs/1433460106?check_suite_focus=true (which errored when it found a bug in my plugin.)

It required one small change in this project https://github.com/timhunt/moodle-plugin-ci/commit/41ef4b614870fc880f46865d9cfd577b6f7f05ad. That is, don't try to install grunt globally with Node. Just install it locally, and the run it using npx. That seems like generially a good idea.

That is the only reason why the familiar `composer create-project` command line has been replaced by separate git checkout and composer install.

Obviously, this needs a lot of cleaning up, but I think that is worth doing.

**Comment by timhunt** [comment]
By the one, someone pointed me at https://ropensci.org/technotes/2020/11/19/moving-away-travis/ today. Some interesting background there.

**Comment by mackensen** [comment]
@timhunt Marvellous, you're a couple of hours ahead of me. I thought https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works was interesting reading as well.

**Comment by timhunt** [comment]
Note: related forum thread https://moodle.org/mod/forum/discuss.php?d=413252

**Comment by timhunt** [comment]
Yay! https://github.com/moodleou/moodle-filter_embedquestion/runs/1433665531?check_suite_focus=true (from https://github.com/moodleou/moodle-filter_embedquestion/blob/github-actions/.github/workflows/ci.yml)

**Comment by mackensen** [comment]
Built on Tim's work to include a matrix with multiple Moodle and PHP versions: https://github.com/LafColITS/moodle-report_roster/runs/1433920311?check_suite_focus=true and https://github.com/LafColITS/moodle-report_roster/blob/actions/.github/workflows/moodle-ci.yml. I'm having some issues getting mariadb/mysqli working properly.

**Comment by kabalin** [comment]
>OK folks, I have sort-of made it work

What else Moodle developer can do on Friday night 😄 Thanks a lot Tim!

Here is my attempt to use `moodle-plugin-ci` with GitHub Actions (also worked on this last night):
https://github.com/kabalin/moodle-enrol_groupsync/pull/1/files

And action run outcome (ignore success status, for debug purposes I override every test exit status with 0):
https://github.com/kabalin/moodle-enrol_groupsync/runs/1435437226?check_suite_focus=true

**Things that we need to consider for now:**
* As GitHub Actions treat non-zero exist status of script run it supposed to ([stops test completely](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference)) and if we need to achieve behaviour similar to Travis, we need some sort of start/finish wrapper #52 
* Alternative to above could be multi job run where each test command is implemented as separate job (and run concurrently if needed) with [`continue-on-error`](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error) flag. Potentially setup part of the job can be separated and re-used, for more info see [Creating dependant jobs](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#creating-dependent-jobs) and [Persisting data between jobs](https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow).

**For future:**
We may consider creating a dedicated [Action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions) for `moodle-plugin-ci`, and list it in the marketplace, so that developer who want to enable CI for their plugin would only need to reference an action in the workflow #53
```
    steps:
      ....
    - name: Setup moodle-plugin-ci
      uses: moodlehq/moodle-plugin-ci@v3
      with:
          database: ${{ matrix.database }}
          branch: ${{ matrix.moodle-version }}
      ...
```

**Comment by kabalin** [comment]
Expanded version inspired by @mackensen and @timhunt examples. This runs MariaDB also as part of the matrix.
https://github.com/kabalin/moodle-enrol_groupsync/pull/2/files

(CI run https://github.com/kabalin/moodle-enrol_groupsync/actions/runs/376107147)

**Comment by timhunt** [comment]
Thanks Folks.

I have been wondering if we really need to keep moodle-plugin-ci in its current form? Could there be a more GitHub-actions-native approach? I mean, what we need to do is:

* Do some git checkouts (which is obviously supported natively)
* Start up some docker containrs for databases, selenium, etc. That is supported natively.
* Run some command-line commands.

Is what we should acutally be doing is to create some action templates like the ones in https://github.com/actions, so we can build our jobs using things like:

```
   uses: moodlehq/action-plugin-phpunit
   with:
     plugin: timhunt/moodle-filter_embedquestion
     plugin-install-path: filter/embedquestion
     moodle-branch: 3.10
     php-version: 7.3
     database: pgsql
```

I am thinking we would just need a few of these:

* action-moodle-phpunit - Run all the PHPunit in core. Used as part of what happens when you submit a fix to the tracker.
* action-moodle-ugprade \
* action-moodle-citest       > These three are the other tests currently run on core fixes. E.g. https://travis-ci.com/github/timhunt/moodle/builds/203079649
* action-moodle-grunt     /
* action-plugin-phpunit - Run PHPUnit on one plugin, like Moodle plugin CI does.
* action-plugin-behat - Run Behat on one plugin, like plugin-ci does.
* action-plugin-codestyle - Run all the different codestyle checks. (Since these are fast, seems sensible to lump them together?)

Would a collection of these be easier to maintain, and make it easier to understand what is going on? (I have never really understood Symphony framework, and I find moodle-plugin-ci obfuscates what is going on.)

Well, I might play with this today.

**Comment by kabalin** [comment]
> I have been wondering if we really need to keep moodle-plugin-ci in its current form? Could there be a more GitHub-actions-native approach?

Yep, that is what I suggested in #53 but with a difference of having one action which will technically replace install step and leave environment in a state ready to carry on running individual commands to test other things.

What you are suggesting also makes sense, but this will mean each test will contain installation step (which is probably OK, but might be not perfect from performance point). Or you are suggesting not using `moodle-plugin-ci` on the background at all for those individual test actions?

IMO `moodle-plugin-ci` should probably stay in the current form because people are using it not only with GitHub, but it should not stop us implementing "actions" or whatever features given CI solution provides having `moodle-plugin-ci` CLI interface on the background.

**Comment by timhunt** [comment]
But: Most of the set-up time is installing the Moodle database. And, the code style tests don't need a DB, and Behat and PHP unit each need a separate database. Better to install those in parallel in two separate jobs. At least, that is my theory.

But, right now, I think it is healthy if different people try different things. Once we see which different options work, we can consolidate on the one we want to maintain.

**Comment by mackensen** [comment]
Playing more with this today; thanks @kabalin for getting to the bottom of the MySQL issue. I have exclusions working on my current branch: https://github.com/LafColITS/moodle-report_roster/blob/actions/.github/workflows/moodle-ci.yml.

@timhunt I find your modular idea intriguing; what I'm trying to figure out is which approach gives us the best "drop in" replacement for travis-ci. Ideally, we can give people a `.github/workflows/moodle-ci.yml-dist` that looks close enough to the current travis file that people don't really have to understand it in order to reimplement.

**Comment by timhunt** [comment]
OK:

* This in my plugin: https://github.com/moodleou/moodle-filter_embedquestion/blob/github-actions/.github/workflows/ci.yml
* And this action: https://github.com/timhunt/moodle-plugin-phpunit/tree/3df7065f2744f436b9a2371321c53461eae02ca6
* Gave this result: https://github.com/moodleou/moodle-filter_embedquestion/runs/1436638776?check_suite_focus=true

Seems you can't use 'uses:' in a sub-action, but then, doing the git checkout in the main script means that you don't have to pass Moodle version and plugin folder into the sub-action, so acutally, quite natural to do it in the main script.

**Comment by timhunt** [comment]
I just learned: `if: always()` is the magic to always run a step even if previous steps failed.

**Comment by mackensen** [comment]
The inclusion/exclusion syntax is a little different too. This:

```
        exclude:
          - {moodle-version: 'MOODLE_37_STABLE', php-versions: '7.4'}
```

...ensures that you never try to test Moodle 3.7 on PHP 7.4 (for all database versions). It seems as though if you want to ''include'' a job, you need a little more:

```
          - {moodle-version: 'MOODLE_37_STABLE', php-versions: '7.1', operating-system: 'ubuntu-latest', database: 'mariadb'}
          - {moodle-version: 'MOODLE_37_STABLE', php-versions: '7.1', operating-system: 'ubuntu-latest', database: 'pgsql'}
```

That adds PHP 7.1 jobs for Moodle 3.7 on both database types. You must specify the operating system. I haven't found a way yet to set defaults.

**Comment by kabalin** [comment]
> But: Most of the set-up time is installing the Moodle database. And, the code style tests don't need a DB, and Behat and PHP unit each need a separate database. Better to install those in parallel in two separate jobs. At least, that is my theory.

Good point. Still might be ok to use `moodle-plugin-ci` for installing in those cases (with `--no-init` option to skip phpunit/behat), also we may add a new option to skip DB install also. In fact, there is a related ticket #36 

> But, right now, I think it is healthy if different people try different things. Once we see which different options work, we can consolidate on the one we want to maintain.

Absolutely! Your action in https://github.com/moodlehq/moodle-plugin-ci/issues/48#issuecomment-731641017 is an interesting example.

**Comment by timhunt** [comment]
Note: Eloy has also been playing with core: https://github.com/stronk7/moodle/actions

**Comment by lushonline** [comment]
I have this working well for my modules, you can see my version for testing MOODLE_35_STABLE thru MOODLE_31_STABLE (inc the temperamental MOODLE_36_STABLE)

[https://github.com/lushonline/moodle-mod_externalcontent/blob/master/.github/workflows/main.yml](https://github.com/lushonline/moodle-mod_externalcontent/blob/master/.github/workflows/main.yml)

**Comment by stronk7** [comment]
Hi,

unless I'm wrong, with the [3.0.4 release](https://moodlehq.github.io/moodle-plugin-ci/CHANGELOG.html) that happened yesterday, this can be finally closed.

We are already providing a reference [gha.dist.yml](https://github.com/moodlehq/moodle-plugin-ci/blob/master/gha.dist.yml) file and it's documented @ [GitHub Actions CI workflow file explained](https://moodlehq.github.io/moodle-plugin-ci/GHAFileExplained.html).

And those were the 2 remaining points in the plan. For your consideration. And thanks to everybody that participated and helped  here!

Ciao :-)

**Comment by kabalin** [comment]
I agree @stronk7, wiki one is not immediately required as all instructions are created here. Closing the issue. Thanks everyone involved.
